DM-4196

jhoblitt opened the pull request.

jonathansick committed:
Create package strucure and setup.py

I'm using the pkg_resources trick for keeping the package's __version__
in sync with setup.py.

For DM-4196.

jonathansick committed:
Setup ltdmason package structure

- Add setup.py
- Add requirements with future, py.test and rumamel.yaml deps
- Package __version__ refers to setup.py's

For DM-4196.

jonathansick committed:
Create Manifest class and YAML schema

Manifest exists as a class wrapper around the YAML-encoded manifest that
serves as buildlsstsw.sh's interface to ltd-mason.

In particular, the manifest provides:

- parsing of YAML text with ruamel.yaml (for better roundtripping of
  YAML comments)
- accessors into the YAML data

A potential schema for the YAML manifest is provided in
tests/demo_science_pipelines_manifest.yaml.

This can be used for live testing, but it is unfortunately specific to
the current lsstsw build.

For DM-4196.

jonathansick committed:
Draft API for the Product class

This API is intended to be used by the build runner script.

For DM-4196.

jonathansick committed:
Implement doc repo checkout in Product

Product.clone_doc_repo() is the first build step in building the docs.
I'm using the sh package since it lets us use the shell git client (and
hence git-lfs as appropriate.

Tests actually checkout the pipelines_doc.git.

For DM-4196.

jonathansick committed:
Implement Product.link_package_repos()

link_package_repos meshes each package's documentation sources into the
product's documentation repository.

Note that _static/ is handled specially;

<pkgname>/doc/_static is link from <doc_repo>/_static/<pkgname>

so that links must be made for all other entities in <pkgname>/doc to be
linked from <doc_repo>/<pkgname>

For DM-4196.

jonathansick committed:
Implement Product.build_sphinx()

Calls sphinx-build using the sh package.

For DM-4196.

jonathansick committed:
Create ltd-mason command line app

This CL app is what buildlsstsw.sh will run to build the docs.

Functionality includes:

- Argument parsing, with the ability to either get the manifest as a
  file or as a stream piped to stdin.
- Run the Product's pipeline of clone -> link -> build.

For DM-4196.

jonathansick committed:
Improve README with installation+usage info

For DM-4196.

jonathansick committed:
Design S3 upload API

The ltdmason.s3upload.upload() function is intended to be a simple to
use synchronization function to push new files to S3, and delete
deprecated files from S3. We're using boto3 to interact with S3.

This commit lays out the structure of the upload function, and puts in
stubs for helper function and classes. These need to be implemented.

The ObjectManager, in particular, is used to find and delete files from
S3 that no longer exist in the source.

This commit builds in the idea of a version_slug/ directory to partition
different versions of the docs in the same bucket.

For DM-4196.

jonathansick committed:
Implement filename and directory listing on S3

Implements ObjectManager.list_filenames_in_directory() and
ObjectManager.list_dirnames_in_directory().

These do all the necessary path manipulation and comparison to detect
with paths exist at the root of a given directory in an S3 bucket. This
is necessary to implement since S3 treats buckets as a flat object store
with directories inferred from object names. That said, Boto3 does
return 'objects' that are directories; the only way to know these objects
are directories is through the trailing / in the key name.

For DM-4196.

jonathansick committed:
Implement S3 object/directory deletion

There are two ObjectManager APIs:

- delete_file
- delete_directory

The delete_directory method uses a more efficient bulk API. There should
probably also be a delete_files method to do the same, but for files in
general.

For DM-4196.

jonathansick committed:
Implement S3 file upload

This turns out to be trivial to implement using boto3's upload_file
method on Objects that automatically does multi-part uploads if
necessary.

For DM-4196.

jonathansick committed:
Install requirements.txt deps from doc repo

Currently this is done within the existing Python environment; ideally a
temporary virtual environment should be created that passes through
global pacakges.

jonathansick committed:
Rename version_slug -> bucket_root

This allows the S3 uploader to be a bit more generic.

jonathansick committed:
list_dirnames_in_directory() uses file listing

Previously ObjectManager.list_dirnames_in_directory() looked for special
'directory objects' whose names ended with '/' to find directories that
existed at the root of a given directory. However these directory
objects aren't generic; they seem to be things that certain clients
such as Transmit.app create, but a general boto3 client will not make.
This commit changes the method logic to find directories from the names
of files.

jonathansick committed:
Add logging to s3upload

The NullHandler is used by default, but one can get log messages via
logging.getLogger('ltdmason.s3upload').level = logging.INFO

jonathansick committed:
Add s3upload_demo script

The point of the demo scripts is to allow for tests against live AWS
services that can't be run as unit tests. s3upload_demo tests file
synchronization to a S3 bucket, including upload and deletion.

jonathansick committed:
PEP8/whitespace cleanup

jonathansick committed:
Allow tests to skip if lsstsw not available

Previously the demo manifest.yaml file in tests/ baked a path to a built
afw. Of course, this will change widely with different machines and
lsstsw installs. A more pragmatic is to have the afw dirname be an
environment variable, $STACK_AFW. pytest will skip any tests that
require the installed afw package.

To make this possible: Manifest resolves environment variables before
shipping Manifest.packages.

For DM-4196.

jonathansick committed:
ltd-mason cli takes --config argument

The config file (~/.ltdmason.yaml) is envisioned as a YAML file with
fields keeper_url and keeper_token containing the URL and an auth token
for a ltd-keeper instance.

For DM-4196.

jonathansick committed:
Add uploader module

The uploaded is responsible for adding a new build to ltd-keeper and
uploading the build to S3. This is where ltd-mason talks to ltd-keeper
via the Keeper's RESTful API.

For DM-4196.

jonathansick committed:
Update manifest spec and Manifest representation

Various updates to the manifest.yaml schema:

- Add product_name (which mirrors the EUPS product name and also the
  ltd-keeper product name).
- Change dirname in package object to dir
- Add optional requester_github_handle field to aide callbacks to the
  person who initiated the build.

The Manifest object now provides accessors to all fields from
manifest.yaml as well.

For DM-4196.

jonathansick committed:
Add logging to Product

This captures information about the product build, including capturing
output from the sh-wrapped command line executions of git and
sphinx-build.

The ltd-mason command line app starts up a logging handler at the INFO
level. This could be made configurable later.

For DM-4196.

jonathansick committed:
Add --build-dir option for debugging

This allows the user to preset the build workspace. This workspace isn't
deleted, so it's easier to debug the build process.

For DM-4196.

jonathansick committed:
Config file now stores keeper username/pass

The ltd-mason command line app now gets the authorization token at
runtime to use with the API.

jonathansick committed:
Add integration test suite with ltd-keeper

These integration test scripts allow you to create a product
(setup_keeper.py) and then use run-ltd-mason.py to trigger a ltd-mason
doc build that uploads the docs via ltd-keeper and S3.

setup_keeper.py creates a product that assumes test AWS resources (the
test bucket and the test Route53 hosted zone). Thus these tests are only
runnable by someone who has lsst sqre AWS credentials.

Also, this requires that the env variable STACK_AFW be set to the path
of the currently built version of afw in the lsstsw.

jonathansick committed:
Add aws_profile configuration

The user can now set the desired AWS configuration profile using the
aws_profile field in config.yaml. This allows a specific AWS IAMs key to
be used for ltd-mason. Internally a boto3 session is made to work with
the given profile.

For DM-4196.

jonathansick committed:
Add doc for integration tests

The integration test environment is a bit complex to set up, so
hopefully this will help. Eventually this README could be imported into
the Sphinx docs.

For DM-4196.

jonathansick committed:
Consolidate integration tests

Removes the aws_demo directory; moves s3upload_demo.py to
integration_tests/ directory.

jhoblitt commented on integration_tests/manifest.yaml:
```diff
@@ -0,0 +1,15 @@
+refs: ['u/jonathansick/DM-4195']
+requester_github_handle: 'jonathansick'
+doc_repo:
+    url: 'https://github.com/lsst-sqre/pipelines_docs.git'
```
We don't have a standard for this but my impression is that 2 space indented yaml is currently more common than 4 space.

jhoblitt commented on ltdmason/manifest.py:
```diff
@@ -0,0 +1,108 @@
+"""Interface for stack manifest YAML files written by buildlsstsw.sh
```
FYI - The `manifest.txt` is written by `lsst_build`'s `prepare` sub-command.

jhoblitt commented on README.rst:
```diff
@@ -2,10 +2,55 @@
+It *must* run from the same Python as the software being documented, however.
+
+::
+
```
I suggest adding a `pyvenv`|`virtualenv` creation and activation example here only because DM seems to have many dev unfamiliar with these tools.

jonathansick replied:
So this is an interesting point because ltd-mason is meant to specifically be installed in the same environment that the Eups product (e.g., lsst_apps) was built in.

@timj do you know the best way to have a Python virtual environment (conda environment or otherwise) layered on top of an lsstsw install?

Is the best way to create an environment, have lsstsw build in that, and then also install/run ltd-mason in that environment? Alternatively I know that virtual environments can incorporate site-packages/, but I don't know mechanism also incorporates packages that weren't actually pip/setup.py install'd into site-packages.

timj replied:
I'm pretty sure that any packages provided by EUPS are going to ignore any virtual environment you have set up. EUPS is a virtual environment that lets you switch between versions at will (so is more powerful than a normal virtual environment). What I'm really saying is that the virtual environment is only controlling packages you build with pip and the like. Anything that is under EUPS control will be distinct so setting up the virtual environment before or after lsstsw runs would seem to be irrelevant.

jonathansick replied:
To summarize our discussions:
- on Jenkins ltd-mason would just be installed and made available in the environment and how that's done would be an operational issue
- virtual environments are for the developer/tester. I like the idea of creating a mock package dataset that can be built easily and portably and using that exclusively for unit tests.

jhoblitt commented on tests/test_manifest.py:
```diff
@@ -0,0 +1,46 @@
+
+@pytest.mark.xfail
+def test_manifest_roundtrip(demo_science_pipelines_manifest):
+    # Can't get perfect round-trip equality of the YAML
```
The assertion is passing.  Does `ruamel` have special semantics for equality testing?

jonathansick replied:
Right, so I couldn't get equality to work in the round trip (comments where shifted etc; it wasn't a character-for-character reproduction of the YAML; i.e., this testing is of string data, not parsed). Hence the `@pytest.mark.xfail` to skip the test.

I'm tempted to drop ruamel in favour of using pyyaml since the round trip feature wasn't as success as I'd hoped.

jonathansick committed:
py.test ignores any venv/ content for discovery

jhoblitt commented on tests/test_manifest.py:
```diff
@@ -0,0 +1,46 @@
+    assert 'afw' in manifest.data['packages']
+    assert 'dir' in manifest.data['packages']['afw']
+    assert 'url' in manifest.data['packages']['afw']
+    assert 'ref' in manifest.data['packages']['afw']
```
It is probably worth testing the value of the keys as well.

jhoblitt commented on tests/test_manifest.py:
```diff
@@ -0,0 +1,46 @@
+    assert 'pipelines_docs' == manifest.doc_repo_name
+    assert 'master' == manifest.doc_repo_ref
+    assert 'https://github.com/lsst-sqre/pipelines_docs.git' \
+        == manifest.doc_repo_url
```
It looks like `build_id` adn `requester_github_handle` are not covered.

jonathansick committed:
MANIFEST.in to package LICENSE, README for PyPI

jhoblitt commented on ltdmason/manifest.py:
```diff
@@ -0,0 +1,108 @@
+    """
+    def __init__(self, f):
+        super().__init__()
+        self.data = ruamel.yaml.load(f, ruamel.yaml.RoundTripLoader)
```
It probably isn't worth the effort at this time but in the future, it might be nice to validate the document at this point.

jonathansick replied:
I think I can actually co-opt http://json-schema.org to also validate YAML structures using https://github.com/Julian/jsonschema

timj commented on tests/test_manifest.py:
```diff
@@ -0,0 +1,46 @@
+install_aliases()
+
+import pkg_resources
+import pytest
```
When did we decide we are using pytest explicitly (rather than using pytest to drive unittest)? :smile:

jonathansick replied:
I take this as part of SQuaRE's carte blanche to ignore the DM Style Guide for SQuaRE services and implement modern practices as it sees fit. :)

timj replied:
It would be nice if SQuaRE had a coding standard then. There was the RFC on Python 3 for SQuaRE and I take PEP8 as a given.

jhoblitt commented on tests/test_product.py:
```diff
@@ -0,0 +1,110 @@
+    manifest = Manifest(demo_science_pipelines_manifest)
+
+    # future doesn't have this?
+    # with tempfile.TemporaryDirectory() as build_dir:
```
`TemporaryDirectory` is being used in `integration_tests/s3upload_demo.py`.  Is this comment stale?

jonathansick replied:
Oops. That test only works on Py 3. Will get rid of that context manager...

timj commented on tests/test_manifest.py:
```diff
@@ -0,0 +1,46 @@
+from __future__ import (division, absolute_import, print_function,
+                        unicode_literals)
+from builtins import *  # NOQA
+from future.standard_library import install_aliases
```
:+1:

jhoblitt commented on tests/test_product.py:
```diff
@@ -0,0 +1,110 @@
+        if os.path.exists(package_static_dir):
+            doc_package_static_dir = os.path.join(product.doc_dir,
+                                                  '_static', str(package_name))
+            assert os.path.exists(doc_package_static_dir)
```
Perhaps verify that it is a symlink ala `islink`?

jhoblitt commented on tests/test_product.py:
```diff
@@ -0,0 +1,110 @@
+            if entity == '_static':
+                continue
+            link_name = os.path.join(target_dir, entity)
+            assert os.path.exists(link_name)
```
Ditto.

jhoblitt commented on integration_tests/manifest.yaml:
```diff
@@ -0,0 +1,15 @@
+        # Set ${STACK_AFW} to the path of the built/installed afw package
+        # in order for tests to run.
+        # dir: '/mnt/stack_docs/lsstsw/stack/Linux64/afw/2015_10.0-14-g7c5ed66'
+        dir: '${STACK_AFW}'
```
It is probably worth testing the variable expansion since this is not a YAML behavior.

jhoblitt commented on integration_tests/s3upload_demo.py:
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python
```
This file is missing execute perms in the repo.

jhoblitt commented on integration_tests/s3upload_demo.py:
```diff
@@ -0,0 +1,101 @@
+    with TemporaryDirectory() as temp_dir:
+        create_test_files(temp_dir, paths)
+
+        s3upload.upload(args.bucketname, 'aws/demo', temp_dir)
```
This invocation signature seems to be wrong.

```
Traceback (most recent call last):
  File "./s3upload_demo.py", line 101, in <module>
    main()
  File "./s3upload_demo.py", line 50, in main
    s3upload.upload(args.bucketname, 'aws/demo', temp_dir)
TypeError: upload() missing 1 required positional argument: 'aws_profile_name'
```

jhoblitt commented on integration_tests/README.rst:
```diff
@@ -0,0 +1,101 @@
+   scons doc
+
+Finally, set the environment variable :envvar:`STACK_AFW` to the directory where the build ``afw`` is located in :file:`lsstsw/stack`.
+
```
Then very finally, run `run-ltd-mason.sh`...?

jonathansick replied:
Yeah, I get to that further down README

jhoblitt commented on integration_tests/README.rst:
```diff
@@ -0,0 +1,101 @@
+   cd lsstsw
+   ./bin/deploy
+   . bin/setup.sh
+   rebuild lsst_apps
```
If the test is only using `afw`, does all of `lsst_apps` need to be built?

jhoblitt commented on integration_tests/README.rst:
```diff
@@ -0,0 +1,101 @@
+   rebuild lsst_apps
+   eups tags --clone bNNNN current
+
+Then get a version of ``afw`` that has Sphinx documentation enabled:
```
instead of building afw twice, how about passing the ref to rebuild?

jonathansick committed:
Indentation of YAML files at 2 spaces

jonathansick committed:
Recommend virtual environments in README

jonathansick committed:
More thoroughly test manifest parsing

- test parsed values directly
- parse all attributes available from Manifest object

jonathansick committed:
Don't use context manager for Py 2.y compat

Once we're Python 3 only we can use the temporary directory context
manager (or create my own).

jonathansick committed:
Test for links specifically

Uses os.path.islink and os.path.lexists to test that what are expected
to be links are in fact links, and that the file behind the link exists.

jonathansick committed:
s3upload_demo provides AWS profile name

jonathansick commented:
I've fixed the low-hanging fruit, but the common theme so far is that I need a better testing strategy that's more portable. Here's my plan:
- [x] Validate YAML Manifest using JSON validation tools; see http://stackoverflow.com/a/22231372
- [x] Create a Mock test doc project rather than using an actual lsstsw installation. This would involve a dummy product documentation repo and some package repos embedded in ltd-mason itself. Then this test could be run from py.test itself since the entire build environment could be setup easily.
  - [x] Create a mock product doc repo on GitHub
  - [x] Embed packages in ltd-mason repo
  - [x] Create py.test tests
- [x] Move S3 tests to unit tests, out of integration tests. I'll do this by adding AWS credentials to a Travis project and having those tests run if travis is available.
  - [x] Create a travis.yml
  - [x] Add a set of [encrypted AWS credentials](https://docs.travis-ci.com/user/encryption-keys/) to travis.yaml
  - [x] Create py.test tests

This would leave the last integration test as one that actually runs the ltd-mason command line app and interacts with an ltd-keeper instance. Perhaps I can move that to unit tests too by mocking the ltd-keeper api using the [Responses package](https://pypi.python.org/pypi/responses).

jonathansick committed:
Validate manifest yaml with a schema

The schema, manifest_schema.yaml in the repo root, is a YAML expression
of JSON schema draft 4. We're able to write down YAML schemas using JSON
schema formats because it's the Python object representations that are
validated, not the text.

See http://spacetelescope.github.io/understanding-json-schema/index.html
for more information on JSON schemas.

Manifest has been updated to run validation on instantiation.

The jsonschema package is a new dependency.

Unit tests now test that validation errors are thrown when required
fields are missing, or when the certain data types are incorrect. This
could be expanded.

For DM-4196.

jonathansick committed:
Remove direct test of Manifest.data

This isn't a useful test since its much more important to test the
Manifest attribute access of manifest data, which is already done.

jonathansick committed:
Embed mock package doc repos for testing

These mock packages, designed to look like LSST Stack packages, can be
used for lightweight end-to-end testing of ltd-mason builds. Since
package repos have always been intended to pre-exist on a local file
system, this is viable for testing. Note that the *product* doc repo is
normally git clone'd during a ltd-mason build. The
github.com/lsst-sqre/mock-doc repo is designed to do this.

For DM-4196.

jonathansick committed:
Run a documentation build as a unit test

Rather than relying on an lsstsw/ to test against, ltd-mason now tests
against a very simple test documentation directory at
github.com/lsst-sqre/mock-doc and the packages are embedded in this repo
under test_data

jonathansick committed:
Remove envar expansion in package dir

This was originally a hack to allow testing of an lsstsw stack; but now
that we have a real mock documentation package there's no need for this
complexity.

jonathansick committed:
Initial travis setup

The sudo=false and install field of .travis.yml is the result of some
playing around with the travis container environment.

jonathansick committed:
Add Sphinx as a dependency

This makes sense because ltd-mason is hard-wired to call sphinx-build on
projects; though it may make more sense to later install the project's
dependencies and get sphinx+extensions that way.

jonathansick committed:
Remove mention of lsstsw dir env var for tests

jonathansick committed:
Add travis badge

jonathansick committed:
Convert s3upload test to run in py.test

- Converted s3upload_demo.py to test_s3upload.py and modified it to run
  under py.test
- Tests require environment variables with AWS credentials and a bucket
  to be used for testing. These are documented in the readme
- Add test key env vars (encrypted) to .travis.yml

jonathansick committed:
uploader and s3uploader pass through aws creds

Previously the authentication concept for ltd-mason was to put
credentials into ~/.aws/credentials under a named profile. The name of
the credential profile was the passed through to create session objects.

Now variables for aws key ID, secret and profile name are passed through
allowing for authentication both with a named profile in
~/.aws/crednetials or through (say) environment variables read by the
CLI or tests.

Also solves an issue where the session was correctly used by the
ObjectManager when doing bulk deletes (of a directory on S3).

jonathansick committed:
Raise errors if the API response includes 'Errors'

This seems to be a better way of detecting error when deleting S3
objects.

jonathansick committed:
Fix s3upload test for Python 2.7

os.makedirs(..., exist_ok=True) not available in 2.7

jonathansick committed:
Use environment variables for ltd-mason config

This makes ltd-mason easier to deploy in product environments (Jenkins,
Travis) especially considering that the configuration includes
credentials

jonathansick committed:
Update run-ltd-mason integration test

- Use new environment variable config
- Use the mock documentation project rather than trying to use an lsstsw
  installation

jonathansick committed:
Package dir path in manifest now absolute

We transform the packages.pkg_name.dir metadata into an absolute path
that is implicitly computed relative to the user's current working
directory. Note this might be different from the manifest's directory
and therefore might introduce some confusion. However at the moment
Manifest doesn't know the manifest file's directory (and sometimes the
Manifest may never have been a file).

jonathansick committed:
Rename packages to alpha and beta

jonathansick committed:
Add Product.package_excludes list for linking

This provides convenient black-list for directories/files like conf.py,
_build, etc that shouldn't be blindly symlinked into the product doc
repo at build time.

jonathansick committed:
Update uploader to revised ltd-keeper api

- remove /v1/
- build/<int>/uploaded returns 200

jonathansick committed:
Fix link in README

jonathansick commented:
I've finished cleaning up the way tests and integration tests are done. The only integration test remaining is the one where a local LTD Keeper instance is needed to test API interaction. Even that test has now been simplified to use a [mock documentation project](https://github.com/lsst-sqre/mock-doc), rather than lsstsw.

The configuration system is now overhauled as well to use environment variables. This will be useful for production environments like Travis and Jenkins.

I still need to do mocked testing of the interactions with the ltd-keeper API to properly test the `ltdmason.upload` module.

jonathansick committed:
Ensure the .git dir is never linked (future proof)

jonathansick committed:
Refactor Keeper api integration steps in upload()

This makes unit testing easier.

jonathansick committed:
Add tests for ltd-keeper API interactions

Uses the responses library to mock requests. Requests is added to
requirements.txt as a testing dependency.

jonathansick committed:
Declare requests as a dependency

jonathansick committed:
Add tests for get_keeper_token

jonathansick committed:
Add tests for read_keeper_credentials()

jonathansick committed:
Test environment variable config for cli

read_aws_credentials() is challenging to test because we not only test
the presence of an environment variable, but also the absence. The
delvar monkeypatch method only removes the monkeypatched var; an actual
environment variable of the same name can still leak through. Instead we
put the logic of omitting variables into a function that replaces
os.getenv.

jonathansick committed:
Document a release procedure

Checking the metadata and README is very useful.

jonathansick commented:
Added mocked testing of the API requests with the `responses` package. :trophy:

jonathansick committed:
Remove nightly python from travis tests

There are third-party dependencies failing so it is likely that this
test is overkill for regular use.

jonathansick committed:
Comment on why os.getenv is patched

Removing an environment variable py the py.test mocking facilities does
not prevent an actual environment variable from appearing in the test.

jonathansick committed:
Ensure that package's doc directory exists

Skips linking that package if there is no doc/ directory

jonathansick committed:
Validate that symlink target exists

Test existence of source before linking; skip otherwise

jonathansick commented:
Summary of code review actions with @jhoblitt
- [x] Change a lot of the `log.info()` calls to `log.debug()` and plan to have more structured information in log.info that can be used for automated metrics.
- [x] Check if `bottom`'s `upload_file` emits meaningful exceptions (otherwise a manual upload validation step needs to be done. _Yep: S3UploadFailedError._
- [x] Check that a package has a `doc` directory, and skip it during the build process entirely if it does not
- [x] When making symlinks, verify that the symlink does point to a file/directory
- [ ] Optionally think about embedding https://github.com/lsst-sqre/mock-doc/ in this repo, and do a subtree git clone.

jonathansick committed:
Fix entity in LICENSE copyright

jonathansick committed:
Add --verbose command line arg for debug logging

The default otherwise is 'info' level logging. During production we
probably want to use verbose logging to pick everything up for later
diagnosis.

More work/thinking is needed to prepare Mason for Panopticon logging
capture (perhaps putting specific log messages into a parseable JSON
or YAML format.

jonathansick committed:
Change most logging messages to DEBUG level

Idea is that plain English messages of workflow status should be treated
as debugging messages. Another level will be used to structured messages
that can be parsed and turned into metrics or alerts with the logging
system.

jonathansick committed:
Keeper build confirmation uses new PATCH method

The Keeper api has changed from confirming an uploading using a POST to
a build_id/uploaded API to a new PATCH method on the build resource
itself.

jonathansick commented:
Updated the upload confirmation API call to Keeper (it now uses a PATCH method) and fixed an issue where the test_uploader.py tests weren't being run
